include/nbl/builtin/hlsl/limits.hlsl

```diff
@@ -0,0 +1,206 @@
+// Copyright (C) 2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_LIMITS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_LIMITS_INCLUDED_
+
+#include <nbl/builtin/hlsl/cpp_compat.hlsl>
+#include <nbl/builtin/hlsl/type_traits.hlsl>
+
+// C++ headers
+#ifndef __HLSL_VERSION
+#include <limits>
+#endif
+
+namespace nbl
+{
+namespace hlsl
+{
+
+#ifdef __HLSL_VERSION
+
+enum float_denorm_style {
+    denorm_indeterminate = -1,
+    denorm_absent        = 0,
+    denorm_present       = 1
+};
+
+enum float_round_style {
+    round_indeterminate       = -1,
+    round_toward_zero         = 0,
+    round_to_nearest          = 1,
+    round_toward_infinity     = 2,
+    round_toward_neg_infinity = 3
+};
+
+template<class T>
+struct numeric_limits;
+
+template<class T> 
+struct num_traits
+{
+    NBL_CONSTEXPR_STATIC_INLINE T MIN            = numeric_limits<T>::is_signed ? (T(1)<<numeric_limits<T>::digits) : T(0);
+    NBL_CONSTEXPR_STATIC_INLINE T MAX            = ~(T(numeric_limits<T>::is_signed)<<numeric_limits<T>::digits);
+    NBL_CONSTEXPR_STATIC_INLINE T DENORM_MIN     = T(0);
+    NBL_CONSTEXPR_STATIC_INLINE T QUIET_NAN      = T(0);
+    NBL_CONSTEXPR_STATIC_INLINE T SIGNALING_NAN  = T(0);
+};
+
+#ifdef NBL_HALF_DEFINED
+template<> 
+struct num_traits<half>
+{
+    NBL_CONSTEXPR_STATIC_INLINE half MIN         = 6.103515e-05;
+    NBL_CONSTEXPR_STATIC_INLINE half MAX         = 65504;
+    NBL_CONSTEXPR_STATIC_INLINE half DENORM_MIN  = 5.96046448e-08;
+};
+#endif
+
+template<> 
+struct num_traits<float32_t>
+{
+    NBL_CONSTEXPR_STATIC_INLINE float32_t MAX           = 3.402823466e+38F;
+    NBL_CONSTEXPR_STATIC_INLINE float32_t MIN           = 1.175494351e-38F;
+    NBL_CONSTEXPR_STATIC_INLINE float32_t DENORM_MIN    = 1.401298464e-45F;
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t  QUIET_NAN     = 0x7FC00000;
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t  SIGNALING_NAN = 0x7FC00001;
+};
+
+template<> 
+struct num_traits<float64_t>
+{
+    NBL_CONSTEXPR_STATIC_INLINE float64_t MAX           = 1.7976931348623158e+308;
+    NBL_CONSTEXPR_STATIC_INLINE float64_t MIN           = 2.2250738585072014e-308;
+    NBL_CONSTEXPR_STATIC_INLINE float64_t DENORM_MIN    = 4.9406564584124654e-324;
+    NBL_CONSTEXPR_STATIC_INLINE uint64_t  QUIET_NAN     = 0x7FF8000000000000ull;
+    NBL_CONSTEXPR_STATIC_INLINE uint64_t  SIGNALING_NAN = 0x7FF0000000000001ull;
+};
+
+template<class T>
+struct numeric_limits
+{
```

Comment:
hmm I'd actually define this in `impl` outside of `__HLSL_VERSION`, simply so that we can add a test full of `static_assert` on the CPU that `nbl::hlsl::impl::numeric_limits==nbl::hlsl::numeric_limits`

P.s. obviously `nbl::hlsl::numeric_limits` would still be an alias for `std::numeric_limits`